### ISSUE TITLE
Add CLI command to list vaults

### ISSUE THREAD

**Issue by adrianbenavides** [build-trust/ockam#3935]
Depends on: https://github.com/build-trust/ockam/issues/3933

Add a new command to list the available vaults. For each vault, show the same information as the `vault show` command.

A new `list` method must be implemented for VaultsState, similar to the NodesState's list method.

This command should prevent the user from deleting the vault if it's being used by a node. You will have to list the nodes, and check for each one its default_vault (`opts.state.nodes.list()` + `node.config.default_vault`).

---

_We love helping new contributors!_
_If you have questions or need help as you work on your first Ockam contribution, please leave a comment on [this discussion](https://github.com/ockam-network/ockam/discussions/1081)._
_If you're looking for other issues to contribute to, checkout [this discussion](https://github.com/build-trust/ockam/discussions/3206) and labels -  https://github.com/build-trust/ockam/labels/good%20first%20issue or https://github.com/build-trust/ockam/labels/help%20wanted_

**Comment by nidnogg** [comment]
I'd like to have a shot at this if there are no takers. 
I noticed the parent issue has an assignee, but if this can be done separately, I don't mind setting up my env here

**Comment by 0x00A5** [comment]
If I am not mistaking, this will be implemented by exposing the `VaultsState::list` method that we introduced in #4106.

**Comment by adrianbenavides** [comment]
> If I am not mistaking, this will be implemented by exposing the `VaultsState::list` method that we introduced in #4106.

Yes, there is an in-progress PR by @hargut adding this command. We'll wait for him in case he wants to finish it.

**Comment by hargut** [comment]
Hi @adrianbenavides,

sorry for the late response to your comments. I've rebased the PR, use the now already available `list()` function, and added the `commands.bats` test.

As I had the impression that we already had one similar comment on another PR and I wanted to dig deeper into the rust language features I've prepared a [patch](https://github.com/build-trust/ockam/commit/2d2044da57659fc6de4821eab734fe757fcc9383) that moves the `cli_state.rs` towards traits with generics and default implementations where possible.
Currently the patch does not provide much added value (features). It unifies the functions and removes duplications where possible. As during the past days there where several changes the patch is right now out-of-sync to the `develop` branch. 

In case it is of interest for you, and/or you have ideas/comments for improvement please let me know. I can prepare a rebased version and enhance it.


Best regards,
Harald

**Comment by adrianbenavides** [comment]
> As I had the impression that we already had one similar comment on another PR and I wanted to dig deeper into the rust language features I've prepared a [patch](https://github.com/build-trust/ockam/commit/2d2044da57659fc6de4821eab734fe757fcc9383) that moves the cli_state.rs towards traits with generics and default implementations where possible.

Can you post the link again? It seems it's broken. I'll take a look at the patch as soon as I can, thanks for taking the time to look into that!

**Comment by hargut** [comment]
Hi @adrianbenavides,

sorry it seems to have been a wrong link, this one should work:
https://github.com/hargut/ockam/commit/2d2044da57659fc6de4821eab734fe757fcc9383

**Comment by adrianbenavides** [comment]
Hey @hargut, I reviewed your patch and there are some good points there. I think we could integrate at least one of the traits, but I'd wait a little bit until we stabilize this code. The `CliState` contents are still in development and adding traits to generalize the behavior of the different components is assuming too much. It's easier to modify if we keep them decoupled, even if there are some parts duplicated.

I'll keep it in mind as we progress, thanks @hargut!

**Comment by hargut** [comment]
Hi @adrianbenavides,

thank you for the feedback.

I fully understand your points and during the learning I've as well thought that it might be easier and more flexible when using free functions for duplicated code parts.


Best regards,
Harald
